src/dom_utils.rs

```diff
@@ -0,0 +1,383 @@
+        Self: Sized;
+
+    /// Clones the node and its entire subtree (including template contents).
+    fn deep_clone(&self) -> Self;
+
+    /// Create a new element, with the given children.
+    fn create_element(name: LocalName) -> ElementBuilder<Self>
+    where
+        Self: Sized;
+
+    /// Create a new text node.
+    fn create_text_node(text: impl Into<StrTendril>) -> Self
+    where
+        Self: Sized;
+}
+
+/// Convenience helper for constructing nodes. Use like:
+///   Handle::create_element(local_name!("a"))
+///       .attribute(&local_name!("href"), "/")
+///       .text("Home")
+///       .build()
+pub struct ElementBuilder<T: NodeHandleExt + Sized> {
+    element: T,
+}
+
+impl<T: NodeHandleExt + Sized> ElementBuilder<T> {
+    pub fn attribute(self, name: &LocalName, value: impl Into<StrTendril>) -> Self {
+        self.element
+            .set_attribute(&QualName::new(None, ns!(), name.clone()), value.into());
+        self
+    }
+
+    pub fn children(self, children: impl Iterator<Item = T>) -> Self {
+        self.element.append_children(children);
+        self
+    }
+
+    pub fn child(self, child: T) -> Self {
+        self.children(std::iter::once(child))
+    }
+
+    pub fn text(self, text: impl Into<StrTendril>) -> Self {
+        self.child(<T as NodeHandleExt>::create_text_node(text))
+    }
+
+    pub fn build(self) -> T {
+        self.element
+    }
+}
+
+/// Recursively visits every DOM node (preorder). Template contents are visited
+/// after children, but there are seldom both.
+pub fn scan_dom<F: FnMut(&Handle)>(handle: &Handle, f: &mut F) {
+    f(handle);
+
+    for child in handle.children.borrow().iter() {
+        scan_dom(child, f);
+    }
+
+    if let NodeData::Element {
+        template_contents: ref tc,
+        ..
+    } = handle.data
+    {
+        if let Some(ref tc_handle) = *tc.borrow() {
+            scan_dom(tc_handle, f);
+        }
+    }
+}
+
+/// Given a <dt> element, find the corresponding <dd> elements.
+///
+/// This is more subtle than you might immediately think, because there can be
+/// multiple <dt> listing various terms with one or more common <dd>
+/// definitions. We need to find the <dt> in the child list, and then skip it
+/// and any other <dt>, before providing the <dd> that follow.
+pub fn dt_descriptions(dt: &Handle) -> Vec<Handle> {
+    assert!(dt.is_html_element(&local_name!("dt")));
+    if let Some(ref dl) = dt
+        .parent_node()
+        .filter(|n| n.is_html_element(&local_name!("dl")))
+    {
+        dl.children
+            .borrow()
+            .iter()
+            .filter(|n| n.is_element())
+            .skip_while(|n| !Rc::ptr_eq(n, dt))
+            .skip_while(|n| n.is_html_element(&local_name!("dt")))
+            .take_while(|n| n.is_html_element(&local_name!("dd")))
+            .cloned()
+            .collect()
+    } else {
+        Vec::new()
+    }
+}
+
+impl NodeHandleExt for Handle {
+    fn parent_node(&self) -> Option<Handle> {
```

Comment:
Can you help me understand the need for this helper? On the surface it seems to just be a more complicated version of the `.parent` getter, but I'm sure I must be missing something? Is there a benefit to setting the parent to a weak version of itself, and returning the strong ref here?

Reply:
`self.parent` is a `Cell<Option<Weak<Node>>>`: https://docs.rs/markup5ever_rcdom/latest/markup5ever_rcdom/struct.Node.html#structfield.parent

See this for a more detailed discussion of interior mutability with cells: https://doc.rust-lang.org/std/cell/index.html#cellt

Since `Weak<>` isn't `Copy`, neither is `Option<Weak<>>`. Thus the only way to get the value of the cell is to move it out of the cell (here, done with `take`).

Once we have that weak handle, we can obtain the strong handle that callers generally want to use (because callers here are assuming that the parent does, in fact, exist). Then, since we took the weak handle out, we need to put it back to avoid changing the state of the node.

So, uh, yes, it's a more complicated version of a getter to satisfy the static and dynamic safety checks that apply to interior mutability cells.